### PR TITLE
Consider "used" or "unused" CNPs / CCNPs as a new metric

### LINKED ISSUE #13338: Consider "used" or "unused" CNPs / CCNPs as a new metric
This could be useful in large-scale environments with many policies. We have seen occurrences of many stale policies ("unused") taking up resources in Cilium, causing memory issues and whatnot. The solution is not to garbage-collect or remove "unused" user applied policies, but rather to surface this information as a metric to an operator. This information will enable them to be proactive. 

I can imagine a metrics dashboard where an operator can see "used" policies vs "unused" policies, and the number of "unused" policies continues to grow, which can be interpreted as a sign that some policies need to be deleted. Ideally, the number of "unused" policies should be steady.

### PR DESCRIPTION
**Pull request by CBVenkataReddy**
Added "used" policy and "unused" policy per-node metrics

Fixes: #13338

Signed-off-by: Venkata Reddy <venkata.reddy@accuknox.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

**Review by errordeveloper — APPROVED** [review]
Thanks very much, this looks good overall :)

File: daemon/cmd/policy.go [hunk 1]
```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
...
 type PolicyReactionEvent struct {
-	wg                *sync.WaitGroup
-	epsToBumpRevision *policy.EndpointSet
-	endpointsToRegen  *policy.EndpointSet
-	newRev            uint64
-	upsertIdentities  map[string]*identity.Identity // deferred CIDR identity upserts, if any
-	releasePrefixes   []*net.IPNet                  // deferred CIDR identity deletes, if any
+	wg                  *sync.WaitGroup
+	epsToBumpRevision   *policy.EndpointSet
+	addendpointsToRegen *policy.EndpointSet
```
Comment (nebril) [inline daemon/cmd/policy.go]: I would prefer for CamelCase to at least be retained here (`addEndpointsToRegen`), ideally to not abbreviate the `added` to `add` too (`addedEndpointsToRegen`).

File: daemon/cmd/policy.go [hunk 2]
```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
...
+	delendpointsToRegen *policy.EndpointSet
```
Comment (nebril) [inline daemon/cmd/policy.go]: Same as above, `deletedEndpointsToRegen` would be preferrable.

File: test/k8sT/Policies.go [hunk 3]
```diff
@@ -1316,6 +1319,312 @@ var _ = SkipDescribeIf(func() bool {
 
 	})
 
+	Context("Basic Policy Metrics Test", func() {
```
Comment (nebril) [inline test/k8sT/Policies.go:1322]: I appreciate the separation between test contexts done here, but I think the tradeoff of readability between test cases and added test runtime is not good here.

I think these new contexts should be merged into one since setup steps them seem to be exactly the same for each of them and just have 3 `It` blocks in the new context.

File: test/k8sT/Policies.go [hunk 4]
```diff
@@ -1316,6 +1319,312 @@ var _ = SkipDescribeIf(func() bool {
...
+		var (
```
Comment (nebril) [inline test/k8sT/Policies.go:1323]: Another thing to consider is whether we want to use our ginkgo based tests for metric testing - this doesn't really integrate anything from k8s side. Maybe metrics should be tested via `cilium-cli`?

**Review by nebril — CHANGES_REQUESTED** [review]
Thanks for the PR!

I have one small stylistic issue and one with the way this change is tested. Details in comments.

**Comment by CBVenkataReddy** [comment]
> Thanks very much, this looks good overall :)

Thank you so much @errordeveloper

File: daemon/cmd/policy.go (hunk 1 above)
Comment (CBVenkataReddy) [reply daemon/cmd/policy.go]: Done

File: daemon/cmd/policy.go (hunk 2 above)
Comment (CBVenkataReddy) [reply daemon/cmd/policy.go]: Done

File: test/k8sT/Policies.go (hunk 4 above)
Comment (CBVenkataReddy) [reply test/k8sT/Policies.go:1323]: Thank you so much @nebril , Yes agree with you.
I shall raise another issue for metric testing in cilium-cli (https://github.com/cilium/cilium-cli/issues) and track from there, kindly suggest.

**Comment by joestringer** [comment]
Please don't forget to follow the instructions in the PR description:

```
Please ensure your pull request adheres to the following guidelines:

 For first time contributors, read Submitting a pull request
 All code is covered by unit and/or runtime tests where feasible.
 All commits contain a well written commit description including a title,
description and a Fixes: #XXX line if the commit addresses a particular
GitHub issue.
 All commits are signed off. See the section Developer’s Certificate of Origin
 Provide a title or release-note blurb suitable for the release notes.
 Thanks for contributing!
Fixes: #issue-number

<!-- Enter the release note text here if needed or remove this section! -->
```

File: pkg/identity/identitymanager/manager.go [hunk 5]
```diff
@@ -93,12 +93,14 @@ func (idm *IdentityManager) add(identity *identity.Identity) {
 			identity: identity,
 			refCount: 1,
 		}
+		identity.ReferenceCount = 1
```
Comment (jrajahalme) [inline pkg/identity/identitymanager/manager.go:96]: This looks like a bug fix for the `ReferenceCount`. I'd like to see this as a separate commit, hopefully with dedicated unit (go test) testing.

File: pkg/endpoint/policy.go [hunk 6]
```diff
@@ -255,6 +255,7 @@ func (e *Endpoint) regeneratePolicy() (retErr error) {
 		e.getLogger().Debug("Forced policy recalculation")
 	}
 
+	repo.UpdatePolicyMetrics(e.SecurityIdentity, true)
```
Comment (jrajahalme) [inline pkg/endpoint/policy.go:258]: This call should not be done on the endpoint level, would be better to do it within `policy.resolvePolicyLocked()` instead.

File: pkg/policy/repository.go [hunk 7]
```diff
@@ -33,6 +30,8 @@ import (
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/policy/api"
 	"github.com/cilium/proxy/go/cilium/api"
+	"sync"
+	"sync/atomic"
```
Comment (jrajahalme) [inline pkg/policy/repository.go:34]: Should not move standard go package imports here.

File: pkg/policy/repository.go [hunk 8]
```diff
@@ -18,9 +18,6 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	"sync"
-	"sync/atomic"
-
```
Comment (jrajahalme) [inline pkg/policy/repository.go:23]: Please keep the empty line after the standard go imports.

File: pkg/policy/distillery.go [hunk 9]
```diff
@@ -70,6 +70,7 @@ func (cache *PolicyCache) lookupOrCreate(identity *identityPkg.Identity, create
 	if create && !ok {
 		cip = newCachedSelectorPolicy(identity, cache.repo.GetSelectorCache())
 		cache.policies[identity.ID] = cip
+
```
Comment (jrajahalme) [inline pkg/policy/distillery.go:73]: Do not add empty line(s) to the end of a code block.

File: pkg/metrics/metrics.go [hunk 10]
```diff
@@ -210,6 +210,14 @@ var (
...
 	Policy = NoOpGauge
 
+	// PolicyUsed
+	// PolicyUsed is the number of used policies loaded into the agent
+	PolicyUsed = NoOpGauge
+
+	// PolicyUnused
+	// PolicyUnused is the number of unused policies loaded into the agent
+	PolicyUnused = NoOpGauge
+
```
Comment (jrajahalme) [inline pkg/metrics/metrics.go:220]: It seems to me that `PolicyUsed` + `PolicyUnused` should always be equal to `Policy`. Do we have confidence that this is the case? Also, if this is the case, do we really need both of them as `PolicyUnused` could always be derived from `Policy` - `PolicyUsed`?

File: daemon/cmd/policy.go [hunk 11]
```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
...
+	if policyOperation == "update" {
+		if deletedEndpointsToRegen.Len() != 0 { // Delete in use Policy
+			metrics.PolicyUsed.Dec()
+		}
+		if deletedEndpointsToRegen.Len() == 0 { // delete dummy Policy
+			metrics.PolicyUnused.Dec()
+		}
+		metrics.PolicyUnused.Inc()
+		reactToRuleUpdates(epsToBumpRevision, deletedEndpointsToRegen, newRev, upsertIdentities, releasePrefixes)
+		reactToRuleUpdates(epsToBumpRevision, addEndpointsToRegen, newRev, upsertIdentities, releasePrefixes)
```
Comment (jrajahalme) [inline daemon/cmd/policy.go:490]: This would bump policy revision twice on each endpoint, and also upsert Identities twice, as well as release prefixes twice.

I'd much rather see the original `reactToRuleUpdates()` in place with combined endpointset, and have the metrics update be a separate operation to avoid making unnecessary changes to the `reactToRuleUpdates()` calls.

File: daemon/cmd/policy.go [hunk 12]
```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
...
+
+func operationHandle(epsToBumpRevision, deletedEndpointsToRegen *policy.EndpointSet, addEndpointsToRegen *policy.EndpointSet, newRev uint64, upsertIdentities map[string]*identity.Identity, releasePrefixes []*net.IPNet, policyOperation string) {
+
+	if policyOperation == "update" {
+		if deletedEndpointsToRegen.Len() != 0 { // Delete in use Policy
+			metrics.PolicyUsed.Dec()
+		}
+		if deletedEndpointsToRegen.Len() == 0 { // delete dummy Policy
+			metrics.PolicyUnused.Dec()
+		}
```
Comment (jrajahalme) [inline daemon/cmd/policy.go:487]: Why the same if condition in both if statements?

Also, I don't see how decrementing metrics once per update achieve accurate counting of policy rules being used or unused?

File: daemon/cmd/policy.go [hunk 13]
```diff
@@ -332,7 +332,9 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
 	// revision.
 	endpointsToBumpRevision := policy.NewEndpointSet(allEndpoints)
 
-	endpointsToRegen := policy.NewEndpointSet(nil)
+	addEndpointsToRegen := policy.NewEndpointSet(nil)
+	deletedEndpointsToRegen := policy.NewEndpointSet(nil)
```
Comment (jrajahalme) [inline daemon/cmd/policy.go:336]: As a single endpoint may select multiple rules I would not collate the two here.

File: daemon/cmd/policy.go [hunk 14]
```diff
@@ -452,20 +458,52 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
...
+			metrics.PolicyUnused.Inc()
+		}
+		reactToRuleUpdates(epsToBumpRevision, addEndpointsToRegen, newRev, upsertIdentities, releasePrefixes)
+	} else { // policyOperation == "Delete"
+		if deletedEndpointsToRegen.Len() != 0 { // Delelete in use Policy
+			metrics.PolicyUsed.Dec()
+		}
+		if deletedEndpointsToRegen.Len() == 0 { // delete dummy Policy
+			metrics.PolicyUnused.Dec()
+		}
```
Comment (jrajahalme) [inline daemon/cmd/policy.go:504]: If conditions seem identical here.

File: test/k8sT/Policies.go [hunk 15]
```diff
@@ -1316,6 +1319,312 @@ var _ = SkipDescribeIf(func() bool {
...
+			}
+
+			totalPolicyCount := policyUsedCount + policyUnusedCount
+
+			if expectsSuccess {
+				Expect(policyCount == totalPolicyCount).Should(Equal(true), "Policy count does not match with PolicyUsed and PolicyUnUsed\n")
+			}
+		}
+
+		It("checks all policies metrics", func() {
```
Comment (jrajahalme) [inline test/k8sT/Policies.go:1403]: Policy metrics would seem like a good candidate for unit testing rather than integration testing. With unit testing we'd get much faster test iteration and could also cover all the changed code paths much easier by mocking rather than creating real deployments.

File: pkg/policy/repository.go [hunk 16]
```diff
@@ -765,3 +764,26 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
...
+
+	for _, v := range matchingRules {
+		tcount := 0
+		for _, value := range v.metadata.IdentitySelected {
+			if value == true {
+				tcount++
+			}
+		}
+		if tcount == 1 && idnty.ID != identity.ReservedIdentityHealth {
+			if isAdd == true && idnty.ReferenceCount == 1 {
```
Comment (jrajahalme) [inline pkg/policy/repository.go:780]: Identity reference count can also be > 1, e.g., when multiple endpoints share the same identity.

File: pkg/identity/identitymanager/manager.go [hunk 17]
```diff
@@ -93,12 +93,14 @@ func (idm *IdentityManager) add(identity *identity.Identity) {
...
 		for o := range idm.observers {
 			o.LocalEndpointIdentityAdded(identity)
 		}
 
 	} else {
 		idMeta.refCount++
+		identity.ReferenceCount = idMeta.refCount
```
Comment (jrajahalme) [inline pkg/identity/identitymanager/manager.go:103]: Not sure how these fixes to `identity.ReferenceCount` play with the reference counting for local identities in `pkg/identity/cache/local.go`. Likely there is no effect as local identities are only used for CIDRs and these are endpoint identities? (@joestringer)

Comment (jrajahalme) [reply pkg/identity/identitymanager/manager.go:103]: If we end up keeping these `identity.ReferenceCount` updates in here, the locking rules to protect against concurrent updates should be documented.

**Review by jrajahalme — CHANGES_REQUESTED** [review]
Overall I fail to see the value of counting unused rules on the agent level, as in any cluster with multiple nodes some rules will remain unused in some nodes while they may be used in other nodes. IMO the usefulness of this approach should be reconsidered, and maybe pivot to the operator based metric instead, even if it is only available when the policy status is enabled.

Furthermore, it remains a question to me how the various increments and decrements of the new metrics balance out to the overall number of rules, as those updates are currently being done from multiple levels in the code.

`rule.metadata.IdentitySelected` map is empty if the rule is not (locally) used in the node. It would be more robust to count the rules in the repository with locally used identities in the keys of `IdentitySelected` map instead of trying to increment and decrement the metrics depending on the kind of update.

**Comment by jrajahalme** [comment]
@christarazi IMO this approach should be taken back to issue #13338 for discussion w.r.t. my comments above.
